### PR TITLE
Refs #33984 -- Added test for creating copies of model instances with inherited m2m fields.

### PR DESCRIPTION
**Pull request by felixxm**
ticket-33984

### CODE REVIEW

File: tests/model_inheritance/tests.py [hunk 1]
```diff
@@ -191,6 +192,20 @@ def b():
...
+            name="Central market", address="944 W. Fullerton"
+        )
+        supplier.customers.set([restaurant])
+        old_customers = supplier.customers.all()
+        supplier.pk = None
+        supplier.id = None
+        supplier._state.adding = True
+        supplier.save()
+        supplier.customers.set(old_customers)
+        self.assertCountEqual(supplier.customers.all(), old_customers)
```
Comment (kezabelle) [inline tests/model_inheritance/tests.py:207]: Initial read, but I guess given the tests are _passing_ here _without_ the proposed fix in #16047, I'm left wondering if it's actually working correctly[^1] - perhaps you already know better?

If the `suppler.customers` manager is cached onto the instance, and the pk is changed, at the point where you compare _new_ and _old_ customers, how sure are you it's doing The Right Thing with regards to the manager's idea of it's core filters? My _gut feeling_ is that it _could_ be, but also _might not be_, leaning towards the latter.

If for the sake of explicitness it was changed to:
```
self.assertCountEqual(Supplier.objects.get(pk=supplier.pk).customers.all(), old_customers)
```
and passed, that'd be a lot clearer [to me, in reviewing it only], even if arbitrarily more complex than strictly necessary.

[^1]: Though tbf I haven't been keeping up with `main`, so maybe the revert has already landed?

File: tests/model_inheritance/tests.py [hunk 2]
```diff
@@ -191,6 +192,20 @@ def b():
...
 
+    def test_create_copy_with_inherited_m2m(self):
+        restaurant = Restaurant.objects.create()
+        supplier = CustomSupplier.objects.create(
+            name="Central market", address="944 W. Fullerton"
+        )
+        supplier.customers.set([restaurant])
+        old_customers = supplier.customers.all()
+        supplier.pk = None
+        supplier.id = None
```
Comment (charettes) [inline tests/model_inheritance/tests.py:203]: This assignment seems redundant with the one above?

Comment (felixxm) [reply tests/model_inheritance/tests.py:203]: I wanted to use [the documented way](https://docs.djangoproject.com/en/4.1/topics/db/queries/#copying-model-instances) of copying instances.

File: tests/model_inheritance/tests.py (hunk 1 above)
Comment (felixxm) [reply tests/model_inheritance/tests.py:208]: > Though tbf I haven't been keeping up with main, so maybe the revert has already landed?

Nope, it's still there.

> If for the sake of explicitness it was changed to:
> 
> ```
> self.assertCountEqual(Supplier.objects.get(pk=supplier.pk).customers.all(), old_customers)
> ```
> 
> and passed, that'd be a lot clearer [to me, in reviewing it only], even if arbitrarily more complex than strictly necessary.

I tried to improved assertions.

**Review by carltongibson — APPROVED** [review]
OK, looks sensible. Thanks @felixxm 🎁

**Pull request merged**